streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
If this is a concern, we could use a global `public final static boolean` variable, that allows the compiler to remove the code completely for production. We just set it to `false` when doing an RC.

`if (ASSERTIONS_ENABLED) assert ...`

Not sure if this needs to be a general discussion. Does it? Why can't we just start doing it,
 maybe just for Streams? (If I can convince you that it's a useful thing to do -- but I am wondering why it would not be useful... ;) )